### ISSUE TITLE
set MemorySwap to twice the size of memory without checking if swap is enabled or not

### ISSUE THREAD

**Issue by dchen1107** [docker/libcontainer#616]
at https://github.com/docker/libcontainer/blob/master/cgroups/fs/memory.go#L50

libcontainer cgroup handling code set MemorySwap to twice the size of memory without checking if swap is enabled or not. We hit the issue kubernetes/issues/8814, had to work around it by explicitly disabling swap. Shouldn't libcontainer handle such case itself like:

if swap not enabled
    cgroup.MemorySwap = -1
elseif cgroup.MemorySwap == 0
    cgroup.MemorySwap = cgroup.Memory*2
else
   ...

**Comment by dchen1107** [comment]
cc/ @vishh @vmarmol

**Comment by vmarmol** [comment]
Yes, if swap is not supported we should not fail setting the implicit limit we derive.

**Comment by vishh** [comment]
The defaulting should happen only if swap is available.

**Comment by hqhq** [comment]
If swap is not enabled, Docker should pass cgroup.MemorySwap=-1 to libcontainer, does k8s use libcontainer without Docker? Or can you give a specific way how we can reproduce this problem?

**Comment by rjnagal** [comment]
k8 uses remote api to create and then start a container. It supplies a hostconfig in both cases. 
For create():

dockerOpts := docker.CreateContainerOptions{
		Name: BuildDockerName(dockerName, container),
		Config: &docker.Config{
			Env:          makeEnvList(opts.Envs),
			ExposedPorts: exposedPorts,
			Hostname:     containerHostname,
			Image:        container.Image,
			// Memory and CPU are set here for older versions of Docker (pre-1.6).
			Memory:     memoryLimit,
			CPUShares:  cpuShares,
			WorkingDir: container.WorkingDir,
			Labels:     labels,
		},
	}

It uses config instead of hostconfig for compat. This one is not verified by daemon. But I guess the one passed in through Start() is the one that will get used:

hc := &docker.HostConfig{
		PortBindings: portBindings,
		Binds:        binds,
		NetworkMode:  netMode,
		IpcMode:      ipcMode,
		// Memory and CPU are set here for newer versions of Docker (1.6+).
		Memory:     memoryLimit,
		CPUShares:  cpuShares,
	}

Start() does have the logic to verify and disable MemorySwap if the file is not present. But it seems like we are not saving the right value.

The original reporter mentioned that it failed with docker 1.6.2, but worked with v1.3.

I'll try to repro this in the morning.

**Comment by vmarmol** [comment]
I think we can agree that the libcontainer API is broken no? :) I didn't ask for swap, yet libcontainer decided to default it to something even though there was no support. Seems like we should not default it in the case where it is not available. How would the client detect that support is not enabled otherwise?

**Comment by dqminh** [comment]
agree with @vmarmol. I think defaulting memsw value to x2 mem value should happen at docker layer and not at libcontainer. Libcontainer should just take the value and set it.

**Comment by rjnagal** [comment]
+1 we should move the logic to docker and remove it from libcontainer.

**Comment by crosbymichael** [comment]
+1 ;)

Sounds good to me

**Comment by dchen1107** [comment]
+1
